llvm/lib/Transforms/Scalar/LICM.cpp

```diff
@@ -2877,6 +2877,12 @@ static bool hoistBOAssociation(Instruction &I, Loop &L,
     if (auto *I = dyn_cast<Instruction>(Inv))
       I->setFastMathFlags(Intersect);
     NewBO->setFastMathFlags(Intersect);
+  } else if (Opcode == Instruction::Or) {
```

Comment:
This block of code seems like something we should have a helper for.  As a follow up, maybe factor something out of other places we do reassociation?

Reply:
Actually, unless I'm missing something, Reassociate.cpp has the same problem of dropping disjoint flags.  Note that there is the existing OverflowTracking helper, which might serve as a starting place for the API.